File: app/src/main/java/com/elementary/tasks/core/app_widgets/calendar/CalendarMonthFactory.java
package com.elementary.tasks.core.app_widgets.calendar;

import android.app.AlarmManager;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Color;
import androidx.annotation.NonNull;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService;

import com.elementary.tasks.R;
import com.elementary.tasks.core.app_widgets.WidgetDataProvider;
import com.elementary.tasks.core.utils.Configs;
import com.elementary.tasks.core.utils.Prefs;
import com.elementary.tasks.core.utils.ThemeUtil;
import com.elementary.tasks.core.utils.TimeUtil;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

import hirondelle.date4j.DateTime;

/**
 * Copyright 2015 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

public class CalendarMonthFactory implements RemoteViewsService.RemoteViewsFactory {

    @NonNull
    private List<DateTime> mDateTimeList = new ArrayList<>();
    @NonNull
    private List<WidgetItem> mPagerData = new ArrayList<>();
    private Context mContext;
    private int mWidgetId;
    private int mDay;
    private int mMonth;
    private int mYear;

    CalendarMonthFactory(Context ctx, Intent intent) {
        mContext = ctx;
        mWidgetId = intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID);
    }

    @Override
    public void onCreate() {
        mDateTimeList.clear();
        mPagerData.clear();
    }

    @Override
    public void onDataSetChanged() {
        mDateTimeList.clear();

        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());

        SharedPreferences sp =
                mContext.getSharedPreferences(CalendarWidgetConfig.CALENDAR_WIDGET_PREF, Context.MODE_PRIVATE);
        int prefsMonth = sp.getInt(CalendarWidgetConfig.CALENDAR_WIDGET_MONTH + mWidgetId, 0);

        mYear = sp.getInt(CalendarWidgetConfig.CALENDAR_WIDGET_YEAR + mWidgetId, calendar.get(Calendar.YEAR));
        mDay = calendar.get(Calendar.DAY_OF_MONTH);
        mMonth = prefsMonth + 1;

        DateTime firstDateOfMonth = new DateTime(mYear, prefsMonth + 1, 1, 0, 0, 0, 0);
        DateTime lastDateOfMonth = firstDateOfMonth.plusDays(firstDateOfMonth.getNumDaysInMonth() - 1);

        int weekdayOfFirstDate = firstDateOfMonth.getWeekDay();
        int startDayOfWeek = Prefs.getInstance(mContext).getStartDay() + 1;

        if (weekdayOfFirstDate < startDayOfWeek) {
            weekdayOfFirstDate += 7;
        }

        while (weekdayOfFirstDate > 0) {
            DateTime dateTime = firstDateOfMonth.minusDays(weekdayOfFirstDate - startDayOfWeek);
            if (!dateTime.lt(firstDateOfMonth)) {
                break;
            }
            mDateTimeList.add(dateTime);
            weekdayOfFirstDate--;
        }
        for (int i = 0; i < lastDateOfMonth.getDay(); i++) {
            mDateTimeList.add(firstDateOfMonth.plusDays(i));
        }
        int endDayOfWeek = startDayOfWeek - 1;
        if (endDayOfWeek == 0) {
            endDayOfWeek = 7;
        }
        if (lastDateOfMonth.getWeekDay() != endDayOfWeek) {
            int i = 1;
            while (true) {
                DateTime nextDay = lastDateOfMonth.plusDays(i);
                mDateTimeList.add(nextDay);
                i++;
                if (nextDay.getWeekDay() == endDayOfWeek) {
                    break;
                }
            }
        }
        int size = mDateTimeList.size();
        int numOfDays = 42 - size;
        DateTime lastDateTime = mDateTimeList.get(size - 1);
        for (int i = 1; i <= numOfDays; i++) {
            DateTime nextDateTime = lastDateTime.plusDays(i);
            mDateTimeList.add(nextDateTime);
        }
        showEvents();
    }

    private void showEvents() {
        Calendar calendar = Calendar.getInstance();
        Prefs sPrefs = Prefs.getInstance(mContext);
        calendar.setTimeInMillis(TimeUtil.getBirthdayTime(sPrefs.getBirthdayTime()));
        int hour = calendar.get(Calendar.HOUR_OF_DAY);
        int minute = calendar.get(Calendar.MINUTE);
        boolean isFeature = sPrefs.isFutureEventEnabled();
        boolean isRemindersEnabled = sPrefs.isRemindersInCalendarEnabled();

        WidgetDataProvider provider = new WidgetDataProvider(mContext);
        provider.setTime(hour, minute);
        if (isRemindersEnabled) {
            provider.setFeature(isFeature);
        }
        provider.fillArray();
        mPagerData.clear();

        calendar.setTimeInMillis(System.currentTimeMillis());
        int currentDay;
        int currentMonth;
        int currentYear;
        int position = 0;
        do {
            currentDay = calendar.get(Calendar.DAY_OF_MONTH);
            currentMonth = calendar.get(Calendar.MONTH);
            currentYear = calendar.get(Calendar.YEAR);
            boolean hasReminders = provider.hasReminder(currentDay, currentMonth, currentYear);
            boolean hasBirthdays = provider.hasBirthday(currentDay, currentMonth);
            mPagerData.add(new WidgetItem(currentDay, currentMonth, currentYear,
                    hasReminders, hasBirthdays));
            position++;
            calendar.setTimeInMillis(calendar.getTimeInMillis() + AlarmManager.INTERVAL_DAY);
        } while (position < Configs.MAX_DAYS_COUNT);
    }

    @Override
    public void onDestroy() {
        mDateTimeList.clear();
        mPagerData.clear();
    }

    @Override
    public int getCount() {
        return mDateTimeList.size();
    }

    @Override
    public RemoteViews getViewAt(int i) {
        SharedPreferences sp = mContext.getSharedPreferences(
                CalendarWidgetConfig.CALENDAR_WIDGET_PREF, Context.MODE_PRIVATE);
        int theme = sp.getInt(CalendarWidgetConfig.CALENDAR_WIDGET_THEME + mWidgetId, 0);
        CalendarTheme calendarTheme = CalendarTheme.getThemes(mContext).get(theme);
        int itemTextColor = calendarTheme.getItemTextColor();
        int rowColor = calendarTheme.getRowColor();
        int reminderM = calendarTheme.getReminderMark();
        int birthdayM = calendarTheme.getBirthdayMark();
        int currentM = calendarTheme.getCurrentMark();
        int prefsMonth = sp.getInt(CalendarWidgetConfig.CALENDAR_WIDGET_MONTH + mWidgetId, 0);
        RemoteViews rView = new RemoteViews(mContext.getPackageName(), R.layout.month_view_grid);

        ThemeUtil cs = ThemeUtil.getInstance(mContext);

        int selDay = mDateTimeList.get(i).getDay();
        int selMonth = mDateTimeList.get(i).getMonth();
        int selYear = mDateTimeList.get(i).getYear();

        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        int realMonth = calendar.get(Calendar.MONTH);
        int realYear = calendar.get(Calendar.YEAR);

        rView.setTextViewText(R.id.textView, String.valueOf(selDay));
        if (selMonth == prefsMonth + 1) {
            rView.setTextColor(R.id.textView, itemTextColor);
        } else {
            rView.setTextColor(R.id.textView, mContext.getResources().getColor(R.color.material_grey));
        }
        rView.setInt(R.id.background, "setBackgroundResource", rowColor);

        rView.setInt(R.id.currentMark, "setBackgroundColor", Color.TRANSPARENT);
        rView.setInt(R.id.reminderMark, "setBackgroundColor", Color.TRANSPARENT);
        rView.setInt(R.id.birthdayMark, "setBackgroundColor", Color.TRANSPARENT);

        if (mPagerData.size() > 0) {
            for (WidgetItem item : mPagerData) {
                int day = item.getDay();
                int month = item.getMonth() + 1;
                int year = item.getYear();
                if (day == selDay && month == selMonth) {
                    if (item.isHasReminders() && year == selYear) {
                        if (reminderM != 0) {
                            rView.setInt(R.id.reminderMark, "setBackgroundResource", reminderM);
                        } else {
                            rView.setInt(R.id.reminderMark, "setBackgroundColor",
                                    mContext.getResources().getColor(cs.colorReminderCalendar()));
                        }
                    } else {
                        rView.setInt(R.id.reminderMark, "setBackgroundColor", Color.TRANSPARENT);
                    }
                    if (item.isHasBirthdays()) {
                        if (birthdayM != 0) {
                            rView.setInt(R.id.birthdayMark, "setBackgroundResource", birthdayM);
                        } else {
                            rView.setInt(R.id.birthdayMark, "setBackgroundColor",
                                    mContext.getResources().getColor(cs.colorBirthdayCalendar()));
                        }
                    } else {
                        rView.setInt(R.id.birthdayMark, "setBackgroundColor", Color.TRANSPARENT);
                    }
                    break;
                }
            }
        }

        if (mDay == selDay && mMonth == selMonth && mYear == realYear && mMonth == realMonth + 1
                && mYear == selYear) {
            if (currentM != 0) {
                rView.setInt(R.id.currentMark, "setBackgroundResource", currentM);
            } else {
                rView.setInt(R.id.currentMark, "setBackgroundColor",
                        mContext.getResources().getColor(cs.colorCurrentCalendar()));
            }
        } else {
            rView.setInt(R.id.currentMark, "setBackgroundColor", Color.TRANSPARENT);
        }

        calendar.setTimeInMillis(System.currentTimeMillis());
        int hour = calendar.get(Calendar.HOUR_OF_DAY);
        int minute = calendar.get(Calendar.MINUTE);
        calendar.set(Calendar.MONTH, selMonth - 1);
        calendar.set(Calendar.DAY_OF_MONTH, selDay);
        calendar.set(Calendar.YEAR, selYear);
        calendar.set(Calendar.HOUR_OF_DAY, hour);
        calendar.set(Calendar.MINUTE, minute);
        long dateMills = calendar.getTimeInMillis();

        Intent fillInIntent = new Intent();
        fillInIntent.putExtra("date", dateMills);
        rView.setOnClickFillInIntent(R.id.textView, fillInIntent);
        return rView;
    }

    @Override
    public RemoteViews getLoadingView() {
        return null;
    }

    @Override
    public int getViewTypeCount() {
        return 1;
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    private static class WidgetItem {
        private int day, month, year;
        private boolean hasReminders, hasBirthdays;

        WidgetItem(int day, int month, int year, boolean hasReminders, boolean hasBirthdays) {
            this.day = day;
            this.month = month;
            this.year = year;
            this.hasReminders = hasReminders;
            this.hasBirthdays = hasBirthdays;
        }

        boolean isHasBirthdays() {
            return hasBirthdays;
        }

        boolean isHasReminders() {
            return hasReminders;
        }

        public int getYear() {
            return year;
        }

        public void setYear(int year) {
            this.year = year;
        }

        public int getMonth() {
            return month;
        }

        public void setMonth(int month) {
            this.month = month;
        }

        public int getDay() {
            return day;
        }

        public void setDay(int day) {
            this.day = day;
        }
    }
}

File: app/src/main/java/com/elementary/tasks/core/app_widgets/calendar/CalendarMonthFactory.kt
package com.elementary.tasks.core.app_widgets.calendar

import android.appwidget.AppWidgetManager
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Build
import android.util.TypedValue
import android.widget.RemoteViews
import android.widget.RemoteViewsService
import androidx.core.content.ContextCompat
import com.elementary.tasks.R
import com.elementary.tasks.core.app_widgets.WidgetDataProvider
import com.elementary.tasks.core.app_widgets.WidgetUtils
import com.elementary.tasks.core.protocol.StartDayOfWeekProtocol
import com.elementary.tasks.core.utils.Configs
import com.elementary.tasks.core.utils.ThemeProvider
import com.elementary.tasks.core.utils.datetime.DateTimeManager
import com.elementary.tasks.core.utils.params.Prefs
import com.elementary.tasks.core.utils.ui.dp2px
import com.elementary.tasks.home.BottomNavActivity
import org.threeten.bp.LocalDate
import org.threeten.bp.LocalDateTime
import org.threeten.bp.LocalTime
import java.util.Calendar

class CalendarMonthFactory(
  intent: Intent,
  private val context: Context,
  private val prefs: Prefs,
  private val widgetDataProvider: WidgetDataProvider,
  private val dateTimeManager: DateTimeManager
) : RemoteViewsService.RemoteViewsFactory {

  private val dateList = ArrayList<LocalDate>()
  private val pagerData = ArrayList<WidgetItem>()
  private val widgetId: Int =
    intent.getIntExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, AppWidgetManager.INVALID_APPWIDGET_ID)
  private var mDay: Int = 0
  private var mMonth: Int = 0
  private var mYear: Int = 0
  private val prefsProvider = CalendarWidgetPrefsProvider(context, widgetId)

  override fun onCreate() {
    dateList.clear()
    pagerData.clear()
  }

  override fun onDataSetChanged() {
    dateList.clear()

    val calendar = Calendar.getInstance()
    calendar.timeInMillis = System.currentTimeMillis()

    val prefsMonth = prefsProvider.getMonth()
    var year = prefsProvider.getYear()
    mDay = calendar.get(Calendar.DAY_OF_MONTH)
    mMonth = prefsMonth + 1

    if (year < 1) {
      year = LocalDate.now().year
    }
    mYear = year

    val firstDateOfMonth = LocalDate.of(year, prefsMonth + 1, 1)
    val lastDateOfMonth = firstDateOfMonth.plusDays(firstDateOfMonth.lengthOfMonth() - 1L)

    var weekdayOfFirstDate = firstDateOfMonth.dayOfWeek.value
    val startDayOfWeek = StartDayOfWeekProtocol(prefs.startDay).getForCalendar()

    if (weekdayOfFirstDate < startDayOfWeek) {
      weekdayOfFirstDate += 7
    }

    while (weekdayOfFirstDate > 0) {
      val dateTime = firstDateOfMonth.minusDays(weekdayOfFirstDate - startDayOfWeek.toLong())
      if (!dateTime.isBefore(firstDateOfMonth)) {
        break
      }
      dateList.add(dateTime)
      weekdayOfFirstDate--
    }
    for (i in 0L until lastDateOfMonth.dayOfMonth) {
      dateList.add(firstDateOfMonth.plusDays(i))
    }
    var endDayOfWeek = startDayOfWeek - 1
    if (endDayOfWeek == 0) {
      endDayOfWeek = 7
    }
    if (lastDateOfMonth.dayOfWeek.value != endDayOfWeek) {
      var i = 1L
      while (true) {
        val nextDay = lastDateOfMonth.plusDays(i)
        dateList.add(nextDay)
        i++
        if (nextDay.dayOfWeek.value == endDayOfWeek) {
          break
        }
      }
    }
    val size = dateList.size
    val numOfDays = 42 - size
    val lastDateTime = dateList[size - 1]
    for (i in 1L..numOfDays) {
      val nextDateTime = lastDateTime.plusDays(i)
      dateList.add(nextDateTime)
    }
    showEvents()
  }

  private fun showEvents() {
    val birthdayTime = dateTimeManager.getBirthdayLocalTime()

    val isRemindersEnabled = prefs.isRemindersInCalendarEnabled

    birthdayTime?.also { widgetDataProvider.setTime(it) }

    if (isRemindersEnabled) {
      widgetDataProvider.setFeature(prefs.isFutureEventEnabled)
    }
    widgetDataProvider.prepare()
    pagerData.clear()

    var dateTime = LocalDateTime.of(LocalDate.now(), LocalTime.now())

    var position = 0
    do {
      val hasReminders =
        widgetDataProvider.hasReminder(dateTime.dayOfMonth, dateTime.monthValue, dateTime.year)
      val hasBirthdays = widgetDataProvider.hasBirthday(dateTime.dayOfMonth, dateTime.monthValue)
      pagerData.add(
        WidgetItem(
          dateTime.dayOfMonth,
          dateTime.monthValue,
          dateTime.year,
          hasReminders,
          hasBirthdays
        )
      )
      position++
      dateTime = dateTime.plusDays(1)
    } while (position < Configs.MAX_DAYS_COUNT)
  }

  override fun onDestroy() {
    dateList.clear()
    pagerData.clear()
  }

  override fun getCount(): Int {
    return dateList.size
  }

  override fun getViewAt(i: Int): RemoteViews {
    val bgColor = prefsProvider.getBackground()
    val textColor = if (WidgetUtils.isDarkBg(bgColor)) {
      ContextCompat.getColor(context, R.color.pureWhite)
    } else {
      ContextCompat.getColor(context, R.color.pureBlack)
    }
    val prefsMonth = prefsProvider.getMonth()
    val rv = RemoteViews(context.packageName, R.layout.list_item_month_grid)

    val selDay = dateList[i].dayOfMonth
    val selMonth = dateList[i].monthValue
    val selYear = dateList[i].year

    val calendar = Calendar.getInstance()
    calendar.timeInMillis = System.currentTimeMillis()
    val realMonth = calendar.get(Calendar.MONTH)
    val realYear = calendar.get(Calendar.YEAR)

    rv.setTextViewText(R.id.textView, selDay.toString())
    if (selMonth == prefsMonth + 1) {
      rv.setTextColor(R.id.textView, textColor)
    } else {
      rv.setTextColor(R.id.textView, ContextCompat.getColor(context, R.color.material_grey))
    }

    rv.setInt(R.id.currentMark, "setBackgroundColor", Color.TRANSPARENT)
    rv.setInt(R.id.reminderMark, "setBackgroundColor", Color.TRANSPARENT)
    rv.setInt(R.id.birthdayMark, "setBackgroundColor", Color.TRANSPARENT)

    val markHeight = prefsProvider.getRowHeightDp() / 3f
    val markHeightPx = context.dp2px(markHeight.toInt())
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
      rv.setViewLayoutHeight(R.id.currentMark, markHeight, TypedValue.COMPLEX_UNIT_DIP)
      rv.setViewLayoutHeight(R.id.reminderMark, markHeight, TypedValue.COMPLEX_UNIT_DIP)
      rv.setViewLayoutHeight(R.id.birthdayMark, markHeight, TypedValue.COMPLEX_UNIT_DIP)
    } else {
      rv.setInt(R.id.currentMark, "setHeight", markHeightPx)
      rv.setInt(R.id.reminderMark, "setHeight", markHeightPx)
      rv.setInt(R.id.birthdayMark, "setHeight", markHeightPx)
    }

    if (pagerData.size > 0) {
      for (item in pagerData) {
        val day = item.day
        val month = item.month + 1
        val year = item.year
        if (day == selDay && month == selMonth) {
          if (item.isHasReminders && year == selYear) {
            rv.setInt(
              R.id.reminderMark, "setBackgroundColor",
              ThemeProvider.colorReminderCalendar(context, prefs)
            )
          } else {
            rv.setInt(R.id.reminderMark, "setBackgroundColor", Color.TRANSPARENT)
          }
          if (item.isHasBirthdays) {
            rv.setInt(
              R.id.birthdayMark, "setBackgroundColor",
              ThemeProvider.colorBirthdayCalendar(context, prefs)
            )
          } else {
            rv.setInt(R.id.birthdayMark, "setBackgroundColor", Color.TRANSPARENT)
          }
          break
        }
      }
    }

    if (mDay == selDay && mMonth == selMonth && mYear == realYear && mMonth == realMonth + 1
      && mYear == selYear
    ) {
      rv.setInt(
        R.id.currentMark,
        "setBackgroundColor",
        ThemeProvider.colorCurrentCalendar(context, prefs)
      )
    } else {
      rv.setInt(R.id.currentMark, "setBackgroundColor", Color.TRANSPARENT)
    }

    calendar.timeInMillis = System.currentTimeMillis()
    val hour = calendar.get(Calendar.HOUR_OF_DAY)
    val minute = calendar.get(Calendar.MINUTE)
    calendar.set(Calendar.MONTH, selMonth - 1)
    calendar.set(Calendar.DAY_OF_MONTH, selDay)
    calendar.set(Calendar.YEAR, selYear)
    calendar.set(Calendar.HOUR_OF_DAY, hour)
    calendar.set(Calendar.MINUTE, minute)
    val dateMills = calendar.timeInMillis

    val fillInIntent = Intent()
    fillInIntent.putExtra("date", dateMills)
    fillInIntent.putExtra(BottomNavActivity.ARG_DEST, BottomNavActivity.Companion.Dest.DAY_VIEW)
    rv.setOnClickFillInIntent(R.id.textView, fillInIntent)
    return rv
  }

  override fun getLoadingView(): RemoteViews? {
    return null
  }

  override fun getViewTypeCount(): Int {
    return 1
  }

  override fun getItemId(position: Int): Long {
    return position.toLong()
  }

  override fun hasStableIds(): Boolean {
    return true
  }

  data class WidgetItem(
    var day: Int,
    var month: Int,
    var year: Int,
    val isHasReminders: Boolean,
    val isHasBirthdays: Boolean
  )
}